pkg/csi/service/common/authmanager.go

```diff
@@ -135,44 +184,237 @@ func GenerateDatastoreIgnoreMapForBlockVolumes(ctx context.Context, vc *vsphere.
+	fsEnabledMap, err := IsFileServiceEnabled(ctx, allvsanDatastoreUrls, vc)
+	if err != nil {
+		log.Errorf("failed to get if file service is enabled on vsan datastores with error %+v", err)
+		return nil, err
+	}
+
+	var dsURLs []string
+	var dsInfos []*cnsvsphere.DatastoreInfo
+	var entities []types.ManagedObjectReference
+	dsURLToInfoMap := make(map[string]*cnsvsphere.DatastoreInfo)
+	for dsURL, dsInfo := range vsanDsURLToInfoMap {
+		if val, ok := fsEnabledMap[dsURL]; ok {
+			if val {
+				// go through all datastores to build entities
+				dsMoRef := dsInfo.Reference()
+				entities = append(entities, dsMoRef)
+				// Get all vSAN datastore URLs and Info which has vSAN FS enabled
+				dsURLs = append(dsURLs, dsURL)
+				dsInfos = append(dsInfos, dsInfo)
+				dsURLToInfoMap[dsURL] = dsInfo
+			}
 		}
 	}
+	log.Debugf("dsURLToInfoMap is %+v", dsURLToInfoMap)
+	return dsURLToInfoMap, nil
+}
 
-	log.Debugf("auth manager: dsURLs %v dsInfos %v", dsURLs, dsInfos)
-	dsIgnoreURLToInfoMap := make(map[string]*cnsvsphere.DatastoreInfo)
+// IsFileServiceEnabled checks if file service is enabled on the specified datastoreUrls.
+func IsFileServiceEnabled(ctx context.Context, datastoreUrls []string, vc *vsphere.VirtualCenter) (map[string]bool, error) {
+	// Compute this map during controller init. Re use the map every other time.
+	log := logger.GetLogger(ctx)
+	err := vc.Connect(ctx)
+	if err != nil {
+		log.Errorf("failed to connect to VirtualCenter. err=%v", err)
+		return nil, err
+	}
+	err = vc.ConnectVsan(ctx)
+	if err != nil {
+		log.Errorf("error occurred while connecting to VSAN, err: %+v", err)
+		return nil, err
+	}
+	// This gets the datastore to file service enabled map for all the vsan datastores.
+	dsToFileServiceEnabledMap, err := getDsToFileServiceEnabledMap(ctx, vc)
+	if err != nil {
+		log.Errorf("failed to query if file service is enabled on vsan datastores or not. error: %+v", err)
+		return nil, err
+	}
+	log.Debugf("dsToFileServiceEnabledMap is %+v", dsToFileServiceEnabledMap)
+	// Now create a map of datastores which are queried in the method.
+	dsToFSEnabledMapToReturn := make(map[string]bool)
+	for _, datastoreURL := range datastoreUrls {
+		if val, ok := dsToFileServiceEnabledMap[datastoreURL]; ok {
+			if !val {
+				msg := fmt.Sprintf("File service is not enabled on the datastore: %s", datastoreURL)
+				log.Debugf(msg)
+				dsToFSEnabledMapToReturn[datastoreURL] = false
+			} else {
+				msg := fmt.Sprintf("File service is enabled on the datastore: %s", datastoreURL)
+				log.Debugf(msg)
+				dsToFSEnabledMapToReturn[datastoreURL] = true
+			}
+		} else {
+			msg := fmt.Sprintf("File service is not enabled on the datastore: %s", datastoreURL)
+			log.Debugf(msg)
+			dsToFSEnabledMapToReturn[datastoreURL] = false
+		}
+	}
+	return dsToFSEnabledMapToReturn, nil
+}
+
+// getDatastoresWithRequiredPriv gets datastores with required priv for CSI user
+func getDatastoresWithRequiredPriv(ctx context.Context, vc *vsphere.VirtualCenter,
```

Comment:
It looks like `getDatastoresWithRequiredPriv` is returning datastores for which the user has priv to create block volumes only. So, could we rename this method to `getDatastoresWithBlockVolumePrivs`